src/frontend/src/optimizer/plan_node/stream_upstream_sink_union.rs

```diff
@@ -0,0 +1,145 @@
+use crate::optimizer::property::{
+    Distribution, FunctionalDependencySet, MonotonicityDerivation, MonotonicityMap,
+    WatermarkColumns, analyze_monotonicity,
+};
+use crate::stream_fragmenter::BuildFragmentGraphState;
+
+#[derive(Debug, Clone, Educe)]
+#[educe(PartialEq, Eq, Hash)]
+struct UpstreamSinkUnionInner {
+    #[educe(PartialEq(ignore), Hash(ignore))]
+    ctx: OptimizerContextRef,
+    schema: Schema,
+    append_only: bool,
+    // `generated_column` is used to generate the `watermark_columns` field and affects the derivation of subsequent
+    // operators. Since the project operator of `generated_column` is actually on the sink-fragment, not on the table,
+    // only expr can be retained here to determine the non-decreasing column.
+    generated_column_exprs: Option<Vec<ExprImpl>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct StreamUpstreamSinkUnion {
+    pub base: PlanBase<Stream>,
+    inner: UpstreamSinkUnionInner,
+}
+
+impl StreamUpstreamSinkUnion {
+    pub fn new(
+        ctx: OptimizerContextRef,
+        schema: &Schema,
+        append_only: bool,
+        generated_column_exprs: Option<Vec<ExprImpl>>,
+    ) -> Self {
+        let inner = UpstreamSinkUnionInner {
+            ctx,
+            schema: schema.clone(),
+            append_only,
+            generated_column_exprs,
+        };
+
+        Self::new_inner(inner)
+    }
+
+    fn new_inner(inner: UpstreamSinkUnionInner) -> Self {
+        let mut out_watermark_columns = WatermarkColumns::new();
+        let mut out_monotonicity_map = MonotonicityMap::new();
+        if let Some(ref generated_column_exprs) = inner.generated_column_exprs {
+            for (expr_idx, expr) in generated_column_exprs.iter().enumerate() {
+                if let MonotonicityDerivation::Inherent(monotonicity) = analyze_monotonicity(expr) {
+                    out_monotonicity_map.insert(expr_idx, monotonicity);
+                    if monotonicity.is_non_decreasing() && !monotonicity.is_constant() {
+                        out_watermark_columns.insert(expr_idx, inner.ctx.next_watermark_group_id());
+                    }
+                }
+            }
+        }
+
+        // TODO(zyx): Maybe some fields are incorrect.
+        let base = PlanBase::new_stream(
+            inner.ctx.clone(),
+            inner.schema.clone(),
+            Some(vec![]), // stream_key
+            FunctionalDependencySet::new(inner.schema.fields().len()),
+            Distribution::SomeShard,
+            if inner.append_only {
+                StreamKind::AppendOnly
+            } else {
+                StreamKind::Retract
+            },
+            false, // emit_on_window_close
+            out_watermark_columns,
+            out_monotonicity_map,
+        );
+
+        Self { base, inner }
+    }
+}
+
+impl Distill for StreamUpstreamSinkUnion {
+    fn distill<'a>(&self) -> XmlNode<'a> {
+        childless_record("StreamUpstreamSinkUnion", vec![])
```

Comment:
I think we should include things like `generated_columns` and other fields we operate on, and leave out no-op fields. We should do so in `verbose mode` at least. Currently the planner test just shows `StreamUpstreamSinkUnion`.